File: 2. Add Two Numbers.py
# Letcode problem: https://leetcode.com/problems/add-two-numbers/
# Level: medium
# You are given two non-empty linked lists representing two non-negative integers. 
# The digits are stored in reverse order, and each of their nodes contains a single digit. Add the two numbers and return the sum as a linked list.

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    def last_node(self):
        node = self
        while node is not None:
            next_node = node.next
            if next_node is not None:
                node = next_node
            else:
                break
        return node
    def nums2node(self, nums):
        first = True
        for num in nums:
            if first == True:
                self.val = num
                self.next = None
            else:
                node =self.last_node()
                node.next = ListNode(num)
    def num2node(self, num):
        node = self.next
        while num > 0:
            digit = num % 10
            # print('num:', num, " digit:", digit)
            if node is None:
                self.val = digit
                node = self
            else:
                node =self.last_node()
                node.next = ListNode(digit)
            num = num // 10
    
    def node2list(self):
        nums = [self.val]
        node = self.next
        while node is not None:
            nums.append(node.val)
            node = node.next
        return nums


class Solution:

    def reverse_number(self, num):
        reversed_num = 0
        while num != 0:
            digit = num % 10
            reversed_num = reversed_num * 10 + digit
            num //= 10
        return reversed_num

    def node2num(self, node):
        last_node = self.last_node(node)
        num = last_node.val
        node_bf = self.node_before(node, last_node)
        while node_bf is not None:
            num = num*10 + node_bf.val
            last_node = node_bf
            node_bf = self.node_before(node, last_node)
        return num
    
    def node_before(self, root , node):
        node_bf = root
        while node_bf.next is not None:
            if node_bf.next == node:
                return node_bf
            node_bf = node_bf.next
        return None
        
    def last_node(self, node):
        last_node = node
        while last_node.next is not None:
            last_node = last_node.next
        return last_node   

    def num2node(self, num):
        node = None
        if num == 0:
            node = ListNode(0)
        while num > 0:
            digit = num % 10
            if node is None:
                node = ListNode(digit)
            else:
                last_node =self.last_node(node)
                last_node.next = ListNode(digit)
            num = num // 10
        return node

    def print_node(self, node):
        nums = [node.val]
        while node.next is not None:
            nums.append(node.next.val)
            node = node.next
        print("node num:", nums)

    def addTwoNumbers(self, l1, l2):
        num1 = self.node2num(l1)
        num2 = self.node2num(l2)
        print('num1:', num1, num2)
        total = num1 + num2
        l3 = self.num2node(total)
        print('Total:', total)
        self.print_node(l3)
        return l3

if __name__ == '__main__':
    head_1 = ListNode(0)
    # nums_11 = 342
    # [0,8,6,5,6,8,3,5,7]
# [6,7,8,0,8,5,8,9,7]
    nums_11 = 753865680
    head_1.num2node(nums_11)
    nums_12 = head_1.node2list()
    # nums_21 = 465
    nums_21 = 798580876
    head_2 = ListNode(0)
    head_2.num2node(nums_21)
    nums_22 = head_2.node2list()
    print('head_1:', nums_11, nums_12)
    print('head_2:', nums_21, nums_22)
    solution = Solution()
    solution.addTwoNumbers(head_1, head_2)

File: 2. Add Two Numbers.go
// Author: Le Anh Tuan (latuannetnam@gmail.com)
// Letcode problem: 2. Add Two Numbers
// Letcode link: https://leetcode.com/problems/add-two-numbers/
// Level: medium
// Topics: Link list, Math, Recursion
// Problem detail:
// You are given two non-empty linked lists representing two non-negative integers.
// The digits are stored in reverse order, and each of their nodes contains a single digit.
// Add the two numbers and return the sum as a linked list.
// You may assume the two numbers do not contain any leading zero, except the number 0 itself.

package main

import "fmt"

func addTwoNumbersWrong(l1 *ListNode, l2 *ListNode) *ListNode {
	// Reverse 2 list
	rl1 := reverseList(l1)
	rl2 := reverseList(l2)

	// Do rl3 = rl1 + rl2
	prevNode1 := rl1
	node1 := rl1
	node2 := rl2
	remain := 0
	for node1 != nil && node2 != nil {
		sum := node1.Val + node2.Val + remain
		newVal := sum % 10
		fmt.Printf("node1:%d + node2:%d = %d ( remain:%d) => newVal:%d, newRemain:%d\n", node1.Val, node2.Val, sum, remain, newVal, sum/10)
		remain = sum / 10
		node1.Val = newVal
		prevNode1 = node1
		node1 = node1.Next
		node2 = node2.Next
	}

	if node1 == nil && node2 != nil {
		// list1 size < list 2 size
		prevNode1.Next = node2
	}

	fmt.Printf("rl3:%v\n", list2Slice(rl1))

	l3 := reverseList(rl1)

	return l3
}

// Runtime: 20 ms, faster than 32.56% of Go online submissions for Add Two Numbers.
// Memory Usage: 4.4 MB, less than 98.68% of Go online submissions for Add Two Numbers.
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	prevNode1 := l1
	node1 := l1
	node2 := l2
	remain := 0
	for node1 != nil && node2 != nil {
		sum := node1.Val + node2.Val + remain
		newVal := sum % 10
		fmt.Printf("node1:%d + node2:%d = %d ( remain:%d) => newVal:%d, newRemain:%d\n", node1.Val, node2.Val, sum, remain, newVal, sum/10)
		remain = sum / 10
		node1.Val = newVal
		prevNode1 = node1
		node1 = node1.Next
		node2 = node2.Next
	}

	fmt.Printf("Last remain1:%d\n", remain)

	if node1 == nil && node2 != nil {
		// list1 size < list 2 size
		fmt.Println("Extend list1 for list2")
		prevNode1.Next = node2
		for node2 != nil {
			sum := node2.Val + remain
			newVal := sum % 10
			fmt.Printf("node2:%d + remain:%d = %d => newVal:%d, newRemain:%d\n", node2.Val, remain, sum, newVal, sum/10)
			remain = sum / 10
			node2.Val = newVal
			prevNode1 = node2
			node2 = node2.Next
		}
	} else if node1 != nil {
		// list1 size > list 2 size
		fmt.Println("Continue list1")
		for node1 != nil {
			sum := node1.Val + remain
			newVal := sum % 10
			fmt.Printf("node1:%d + remain:%d = %d => newVal:%d, newRemain:%d\n", node1.Val, remain, sum, newVal, sum/10)
			remain = sum / 10
			node1.Val = newVal
			prevNode1 = node1
			node1 = node1.Next
		}
	}

	fmt.Printf("Last remain2:%d\n", remain)

	if remain > 0 {
		fmt.Println("Extend list1 for remain")
		node2 = &ListNode{
			Val: remain,
		}
		prevNode1.Next = node2
	}

	return l1
}

func mainAddTwoNumbers() {
	arr1 := []int{2, 4, 3}
	arr2 := []int{5, 6, 4}

	arr2 = []int{9, 9, 9, 9, 9, 9, 9}
	arr1 = []int{9, 9, 9, 9}

	arr2 = []int{0}
	arr1 = []int{0}

	l1 := slice2List(arr1)
	l2 := slice2List(arr2)
	fmt.Printf("l1:%v l2:%v\n", arr1, arr2)
	l3 := addTwoNumbers(l1, l2)
	fmt.Printf("l1+l2=%v\n", list2Slice(l3))

}
